lab-steve/route/hops-router.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const Router = require('express').Router;
+const jsonParser = require('body-parser').json();
+const Hop = require('../model/hops.js');
+
+let hopsRouter = module.exports = new Router();
+
+hopsRouter.post('/api/hops', jsonParser, (req, res, next) => {
+  console.log('POST /api/hops');
+  new Hop(req.body)
+    .save()
+    .then(hop => res.status(201).json(hop))
+    .catch(next);
+});
+
+hopsRouter.get('/api/hops', (req, res, next) => {
+  console.log('GET /api/hops');
+  Hop.find()
+    .then(hops => hops.map(hop => hop._id))
+    .then(ids => res.status(200).json(ids))
+    .catch(next);
+});
+
+hopsRouter.get('/api/hops/:id', (req, res, next) => {
+  console.log('GET /api/hops/:id');
+  Hop.findById(req.params.id)
+    .then(hop => res.status(200).json(hop))
+    .catch(next);
+});
+
+hopsRouter.put('/api/hops/:id', jsonParser, (req, res, next) => {
+  console.log('PUT /api/hops:id');
+
+  if (Object.keys(req.body).length === 0) {
+    Hop.findById(req.params.id)
+      .then(hop => res.status(400).json(hop))
+      .catch(next);
+  }
```

Comment:
Good catch on the handling of a bad request. However, you may consider not even invoking the mongoose method to query the database if you know that your request doesn't have any information on it to update the resource.
(conditional) => return res.status(400);